riot/lib/SSL.ml

```diff
@@ -0,0 +1,277 @@
+(******************************************************************************************
+
+  The Tls_unix below was ported from  `ocaml-tls`, its `eio` subpackage, specifically from:
+    * https://github.com/mirleft/ocaml-tls/blob/main/eio/tls_eio.ml
+    * https://github.com/mirleft/ocaml-tls/blob/main/eio/x509_eio.ml
+
+  under this license:
+
+    Copyright (c) 2014, David Kaloper and Hannes Mehnert
+    All rights reserved.
+
+    Redistribution and use in source and binary forms, with or without modification,
+    are permitted provided that the following conditions are met:
+
+    * Redistributions of source code must retain the above copyright notice, this
+      list of conditions and the following disclaimer.
+
+    * Redistributions in binary form must reproduce the above copyright notice, this
+      list of conditions and the following disclaimer in the documentation and/or
+      other materials provided with the distribution.
+
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+    ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+    WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+    DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+    ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+    (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+    ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+    (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+    SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+  *******************************************************************************************)
+
+open Logger.Make (struct
+  let namespace = [ "riot"; "net"; "ssl" ]
+end)
+
+module IO = Rio
+
+let ( let* ) = Result.bind
+
+type 'src t = {
+  writer : 'src IO.Writer.t;
+  reader : 'src IO.Reader.t;
+  mutable state : [ `Active of Tls.Engine.state | `Eof | `Error of exn ];
+  mutable linger : string option;
+  recv_buf : bytes;
+}
+
+exception Tls_alert of Tls.Packet.alert_type
+exception Tls_failure of Tls.Engine.failure
+
+module Tls_unix = struct
+  exception Read_error of Rio.io_error
+  exception Write_error of Rio.io_error
+
+  let err_to_str err = Format.asprintf "%a" Rio.pp_err err
+
+  let read_t t dst =
+    let src = IO.Bytes.with_capacity (Bytes.length dst) in
+    match IO.read t.reader src with
```

Comment:
We are reading into a temporary buffer (`src`) that we then copy back into `dst` – `dst` should already be of type `bytes` here, so we should be able to call `IO.read t.reader dst` and avoid the extra `blit` below.